### PR TITLE
[MHA] Implement MIOPEN_BACKEND_POINTWISE_DESCRIPTOR

### PR DESCRIPTION
**Pull request by scerzh**
Issue [#2839](https://github.com/ROCm/MIOpen/issues/2839)
The following tasks were implemented.

MIOPEN_BACKEND_POINTWISE_DESCRIPTOR
- [x] Builder and tests
- [x] C functions and tests

C API -> BackendPointwiseDescriptor -> PointwiseBuilder -> Pointwise

BackendPointwiseDescriptor, a polymorphic descriptor class, is utilized by C-API functions to handle create/set/finalize/get/execute requests.

Builder pattern is used for set and finalize API requests.

Pointwise class holds parameters for a pointwise operation (which is created separately).

### CODE REVIEW

**Comment by scerzh** [comment]
@atamazov , you are welcome to review.
@DrizztDoUrden , could you review our new enum in miopen.h, please?

urgency_blocker is set as further development for [#2839](https://github.com/ROCm/MIOpen/issues/2839) is blocked by this PR

**Comment by junliume** [comment]
@scerzh please update the feature branch by resolving the conflicts, also the latest CI seems to be having issues.

**Comment by scerzh** [comment]
@junliume @apwojcik  all CI jobs succeeded except Windows, which logs show an SSL/connection error:
```
[2024-04-01T17:58:07.205Z] Downloading http://sourceware.org/pub/bzip2/bzip2-1.0.8.tar.gz
[2024-04-01T17:58:07.205Z] 
[2024-04-01T17:58:07.524Z] Unexpected error: <class 'OSError'>
[2024-04-01T17:58:07.524Z] [Errno socket error] [Errno socket error] [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)
[2024-04-01T17:58:07.524Z] Failed to build package bzip2@1.0.8
```
This type of issue doesn't seem to be caused by the PR.

File: include/miopen/miopen.h
Comment (DrizztDoUrden) [inline include/miopen/miopen.h:6146]: [suggestion] It's easier to read large comments when they take full line, so I recommend placing them before the enum element rather than at the end of the line

File: src/graphapi/pointwise.cpp
Comment (DrizztDoUrden) [inline src/graphapi/pointwise.cpp:50]: I am not sure `miopenStatusNotInitialized` is a good error code here. `miopenStatusInvalidValue` is probably a proper one here.

**Review by DrizztDoUrden — APPROVED** [review]
Minor suggestions, nothing blocking

File: src/graphapi/pointwise.cpp
Comment (scerzh) [reply src/graphapi/pointwise.cpp:50]: Agree with you, but I decided to return the same error as described in [CUDNN Graph API docs](https://docs.nvidia.com/deeplearning/cudnn/api/cudnn-graph-library.html#cudnnbackendsetattribute). Do you think we should deviate from it here?

Comment (scerzh) [reply src/graphapi/pointwise.cpp:50]: I suggest we should not unless we have a reason for that.

File: include/miopen/miopen.h
Comment (scerzh) [reply include/miopen/miopen.h:6146]: Agree. If Jun insists on overcoming windows build errors before merging, I will fix comments in this PR, otherwise in the next one.

File: src/graphapi/graphapi.cpp
Comment (amberhassaan) [inline src/graphapi/graphapi.cpp:148]: you may want to refactor it into a single function that works for all descriptor types. Less repetition that way.

**Review by amberhassaan — APPROVED** [review]
Looks good.

File: src/graphapi/pointwise.cpp
Comment (DrizztDoUrden) [reply src/graphapi/pointwise.cpp:50]: It is misleading, but I guess it would be OK to keep the same one they have

File: src/graphapi/graphapi.cpp
Comment (junliume) [reply src/graphapi/graphapi.cpp:148]: Let's take this into the next refinement PR :)

**Review by junliume — APPROVED** [review]

File: include/miopen/miopen.h
Comment (scerzh) [reply include/miopen/miopen.h:6146]: Done in https://github.com/ROCm/MIOpen/pull/2792

File: src/graphapi/graphapi.cpp
Comment (scerzh) [reply src/graphapi/graphapi.cpp:148]: Done in https://github.com/ROCm/MIOpen/pull/2792

File: test/gtest/graphapi_gtest_common.hpp
Comment (atamazov) [inline test/gtest/graphapi_gtest_common.hpp]: @scerzh Yes, we should never use UL and L in _integer_ literals. Static casts or ULL/LL are fine.

Comment (scerzh) [reply test/gtest/graphapi_gtest_common.hpp]: @atamazov what about L or F in floating point literals?

**Review by atamazov — COMMENTED** [review]
🟢 LGTM after all

File: test/gtest/graphapi_gtest_common.hpp
Comment (atamazov) [reply test/gtest/graphapi_gtest_common.hpp]: @scerzh Word "integer" added. Good catch, thanks))

**Pull request merged**
